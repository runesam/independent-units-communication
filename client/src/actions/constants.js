const SOCKET_INIT = 'SOCKET_INIT';
const SOCKET_CLOSE = 'SOCKET_CLOSE';

const PLAYER_INVITE = 'PLAYER_REMOVE';
const PLAYERS_UPDATE = 'PLAYERS_UPDATE';

const INVITATION_RECEIVED = 'INVITATION_RECEIVED';
const INVITATION_ACCEPTED = 'INVITATION_ACCEPTED';
const INVITATION_REJECTED = 'INVITATION_REJECTED';

const PLAYER_INVITE_REJECTED = 'PLAYER_INVITE_REJECTED';
const PLAYER_INVITE_ACCEPTED = 'PLAYER_INVITE_ACCEPTED';

const GAME_INIT = 'GAME_INIT';
const NEXT_MOVE = 'NEXT_MOVE';
const GAME_INITIATED = 'GAME_INITIATED';
const RECEIVE_NEXT_MOVE = 'RECEIVE_NEXT_MOVE';

export {
    SOCKET_INIT,
    SOCKET_CLOSE,

    PLAYERS_UPDATE,
    PLAYER_INVITE,

    INVITATION_RECEIVED,
    INVITATION_ACCEPTED,
    INVITATION_REJECTED,

    PLAYER_INVITE_ACCEPTED,
    PLAYER_INVITE_REJECTED,

    GAME_INIT,
    NEXT_MOVE,
    GAME_INITIATED,
    RECEIVE_NEXT_MOVE,
};
