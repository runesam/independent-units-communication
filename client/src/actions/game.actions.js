import {
    GAME_INIT,
    NEXT_MOVE,
    GAME_INITIATED,
    RECEIVE_NEXT_MOVE,
} from './constants';

const gameInit = () => ({
    type: GAME_INIT,
});

const nextMove = payload => ({
    type: NEXT_MOVE,
    payload,
});

const gameInitiated = payload => ({
    type: GAME_INITIATED,
    payload,
});

const receiveNextMove = payload => ({
    type: RECEIVE_NEXT_MOVE,
    payload,
});

export {
    nextMove,
    gameInit,
    gameInitiated,
    receiveNextMove,
};
