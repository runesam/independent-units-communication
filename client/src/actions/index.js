export * from './constants';
export * from './socket.actions';
export * from './players.actions';
export * from './game.actions';
