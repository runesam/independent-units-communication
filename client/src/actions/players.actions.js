import {
    PLAYER_INVITE,
    PLAYERS_UPDATE,
    INVITATION_RECEIVED,
    INVITATION_ACCEPTED,
    INVITATION_REJECTED,
    PLAYER_INVITE_ACCEPTED,
    PLAYER_INVITE_REJECTED,
} from '.';

const updatePlayers = payload => ({
    type: PLAYERS_UPDATE,
    payload,
});

const invitePlayer = payload => ({
    type: PLAYER_INVITE,
    payload,
});

const invitePlayerAccepted = payload => ({
    type: PLAYER_INVITE_ACCEPTED,
    payload,
});

const invitePlayerRejected = payload => ({
   type: PLAYER_INVITE_REJECTED,
   payload,
});

const invitationReceived = payload => ({
    type: INVITATION_RECEIVED,
    payload,
});

const acceptInvitation = payload => ({
    type: INVITATION_ACCEPTED,
    payload,
});

const rejectInvitation = payload => ({
    type: INVITATION_REJECTED,
    payload,
});

export {
    invitePlayer,
    updatePlayers,

    invitationReceived,
    invitePlayerAccepted,
    invitePlayerRejected,

    acceptInvitation,
    rejectInvitation,
};
