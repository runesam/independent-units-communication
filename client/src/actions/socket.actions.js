import {
    SOCKET_INIT,
    SOCKET_CLOSE,
} from '.';

const socketInit = () => ({
    type: SOCKET_INIT,
});

const socketClose = () => ({
    type: SOCKET_CLOSE,
});

export {
    socketInit,
    socketClose,
};
