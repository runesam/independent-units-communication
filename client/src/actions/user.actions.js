import {
    LOGIN_START,
    LOGIN_SUCCESS,
    LOGIN_FAILURE,
    LOGOUT_START,
    LOGOUT_SUCCESS,
    LOGOUT_FAILURE,
} from './constants';

const login = (username, password) => ({
    type: LOGIN_START,
    payload: { username, password },
});

const loginFailure = reason => ({
    type: LOGIN_FAILURE,
    payload: { reason },
});

const loginSuccess = response => ({
    type: LOGIN_SUCCESS,
    payload: { response },
});

const logout = () => ({
    type: LOGOUT_START,
});

const logoutSuccess = () => ({
    type: LOGOUT_SUCCESS,
});

const logoutFailure = reason => ({
    type: LOGOUT_FAILURE,
    payload: { reason },
});


export {
    login,
    loginFailure,
    loginSuccess,
    logout,
    logoutSuccess,
    logoutFailure,
};
