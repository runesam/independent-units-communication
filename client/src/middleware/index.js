import { applyMiddleware } from 'redux';
import { createBrowserHistory } from 'history';
import { routerMiddleware } from 'connected-react-router';

import mainMiddleware from './main.middleware';
import userMiddleware from './user.middleware';

export const history = createBrowserHistory();
const middleware = [
    mainMiddleware,
    userMiddleware,
    routerMiddleware(history),
];

export default applyMiddleware(...middleware);
