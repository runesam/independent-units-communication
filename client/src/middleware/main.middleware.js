import { push } from 'connected-react-router';

import {
    GAME_INIT,
    NEXT_MOVE,
    SOCKET_INIT,
    PLAYER_INVITE,
    PLAYERS_UPDATE,
    INVITATION_REJECTED,
    INVITATION_ACCEPTED,
    PLAYER_INVITE_ACCEPTED,

    gameInitiated,
} from '../actions';

import socket from '../module/socket';

function gameInit(store) {
    const {
        invite,
        players,
        invitation,
    } = store.getState();

    if (Object.keys(invitation).length) {
        const { number, from: { id } } = invitation;

        return {
            id,
            who: 'pong',
            number: parseInt(number, 10),
            username: players.find(player => player.id === id).username,
        };
    }

    const id = Object.keys(invite).find(item => invite[item].status === 'accepted');
    const { number } = invite[id] || {};
    const player = players.find(item => item.id === id);

    return {
        id,
        who: 'ping',
        username: player.username,
        number: parseInt(number, 10),
    };
}

export default store => next => (action) => {
    const { type, payload } = action;
    switch (type) {
        case SOCKET_INIT:
            socket.init(store);
            next(action);
            break;
        case PLAYER_INVITE:
            socket.invitePlayer(payload);
            next(action);
            break;
        case PLAYER_INVITE_ACCEPTED:
            next(action);
            store.dispatch(push('/game'));
            break;
        case INVITATION_REJECTED:
            socket.rejectInvitation(payload);
            next(action);
            break;
        case NEXT_MOVE:
            socket.sendNextMove(payload);
            next(action);
            break;
        case GAME_INIT:
            store.dispatch(gameInitiated(gameInit(store)));
            next(action);
            break;
        case PLAYERS_UPDATE:
            if (payload.length === 0) {
                store.dispatch(push('/'));
            }
            next(action);
            break;
        case INVITATION_ACCEPTED:
            socket.acceptInvitation(payload);
            store.dispatch(push('/game'));
            next(action);
            break;
        default: next(action);
    }
};
