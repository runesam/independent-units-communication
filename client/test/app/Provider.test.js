import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Cookies from 'js-cookie';

import { history } from 'middleware';
import { loginSuccess } from 'actions';
import Provider from 'app/provider';

describe('<Provider />', () => {
    const calledWith = [];
    const store = {
        dispatch: jest.fn((arg) => {
            calledWith.push(arg);
            return arg;
        }),
        subscribe: jest.fn(),
        getState: jest.fn(),
    };

    describe('behaviour when there is no auth in cookies', () => {
        it('doesn\'t dispatches loginSuccess action', () => {
            shallow(
                <Provider store={store} history={history}>
                    <span>app</span>
                </Provider>,
            );
            expect(store.dispatch).toBeCalledTimes(0);
        });
    });

    describe('behaviour when there is auth in cookies', () => {
        it('dispatches loginSuccess action', () => {
            // mock cookie
            jest.mock('js-cookie');
            Cookies.getJSON = () => ({ token: 'token' });

            const action = loginSuccess({ data: { token: 'token' } });
            shallow(
                <Provider store={store} history={history}>
                    <span>app</span>
                </Provider>,
            );
            expect(store.dispatch).toBeCalledTimes(1);
            expect(store.dispatch).toBeCalledWith(action);
        });
    });

    it('matches the snapshot', () => {
        const wrapper = shallow(
            <Provider store={store} history={history}>
                <span>app</span>
            </Provider>,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
