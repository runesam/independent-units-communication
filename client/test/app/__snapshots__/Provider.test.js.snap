// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Provider /> matches the snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "payload": Object {
                "response": Object {
                  "data": Object {
                    "token": "token",
                  },
                },
              },
              "type": "LOGIN_SUCCESS",
            },
          ],
          Array [
            Object {
              "payload": Object {
                "response": Object {
                  "data": Object {
                    "token": "token",
                  },
                },
              },
              "type": "LOGIN_SUCCESS",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Object {
              "payload": Object {
                "response": Object {
                  "data": Object {
                    "token": "token",
                  },
                },
              },
              "type": "LOGIN_SUCCESS",
            },
          },
          Object {
            "type": "return",
            "value": Object {
              "payload": Object {
                "response": Object {
                  "data": Object {
                    "token": "token",
                  },
                },
              },
              "type": "LOGIN_SUCCESS",
            },
          },
        ],
      },
      "getState": [MockFunction],
      "subscribe": [MockFunction],
    }
  }
>
  <Theme>
    <Connect(ConnectedRouterWithContext)
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <span>
        app
      </span>
    </Connect(ConnectedRouterWithContext)>
  </Theme>
</Provider>
`;
